arbor/cable_cell.cpp

```diff
@@ -168,24 +171,32 @@ const cable_cell_region_map& cable_cell::region_assignments() const {
     return impl_->region_map;
 }
 
-// Forward paint methods to implementation class.
-
-#define FWD_PAINT(proptype)\
-void cable_cell::paint(const region& target, proptype prop) {\
-    impl_->paint(target, prop);\
+// Forward paint method to implementation class.
+void cable_cell::paint(const region& target, paintable prop) {
+    std::visit(
+            [this, &target] (auto&& p) {impl_->paint(target, p);},
+            prop);
+    decor_.paintings.push_back({target, prop});
 }
-ARB_PP_FOREACH(FWD_PAINT,\
-    mechanism_desc, init_membrane_potential, axial_resistivity,\
-    temperature_K, membrane_capacitance, init_int_concentration,
-    init_ext_concentration, init_reversal_potential)
 
-// Forward place methods to implementation class.
+// Forward place method to implementation class.
+lid_range cable_cell::place(const locset& target, placeable prop) {
+    auto lids = std::visit(
+            [this, &target] (auto&& p) -> lid_range {return impl_->place(target, p);},
+            prop);
 
-#define FWD_PLACE(proptype)\
-lid_range cable_cell::place(const locset& target, proptype prop) {\
-    return impl_->place(target, prop);\
+    decor_.placements.push_back({target, prop});
+    return lids;
+}
+
+void cable_cell::decorate(const decor& d) {
+    for (auto& x: d.paintings) {
+        paint(x.first, x.second);
+    }
+    for (auto& x: d.placements) {
+        place(x.first, x.second);
+    }
+    decor_.defaults = d.defaults;
 }
```

Comment:
Is it always the case that `decor_` will end up with a copy of the parameter `d`?

The current implementation with `decorate` is a bit of a mix between hidden and exposed functionality. It's not clear that the private `paint` and `place` methods are only called during initialization, nor that `decor_` remains constant over the lifetime of the cable cell. All the remaining cable cell functionality lives in the `cable_cell_impl` class.

To clarify these conditions, I'd prefer if:
* `decor_` were a const member that is initialized in the `cable_cell` constructor;
* the constructor calls a private `init` method that does the work of creating the `cable_cell_impl` object, which in turn takes the `decor` as a parameter to the constructor;
* and the `cable_cell_impl` constructor does what `decorate`, `paint`, and `place` do here.

To make the dependency and constness even more clear, one could have in the header file:
```
struct cable_cell_impl;
std::unique_ptr<cable_cell_impl, void (*)(cable_cell_impl*)> make_cable_cell_impl(const arb::morphology&, const label_dict&, const decor&);
```
and with `cable_cell` constructor:
```
cable_cell(const class morphology& m, const label_dict& l, const decor& d):
    decor_(d), impl_(make_cable_cell_impl_(m, l, d)) {}
```

This and a const `decor_` member makes it clear that `decor_` is just there to answer the `decorations()` query, and that everything else is the responsibility of the pimpl class.

Reply:
Seems reasonable. I will have a look at a time estimate for the work, and whether it can go in a separate PR or be merged here.